controller/api/destination/watcher/endpoints_watcher.go

```diff
@@ -1146,8 +1167,9 @@ func (pp *portPublisher) resolveTargetPort(subset corev1.EndpointSubset) Port {
 func (pp *portPublisher) updateLocalTrafficPolicy(localTrafficPolicy bool) {
 	pp.localTrafficPolicy = localTrafficPolicy
 	pp.addresses.LocalTrafficPolicy = localTrafficPolicy
+	addresses := pp.addresses.shallowCopy()
 	for _, listener := range pp.listeners {
-		listener.Add(pp.addresses)
+		listener.Add(addresses)
```

Comment:
this is passing one shared copy of the map to all the listeners.  as it is, this is okay since listeners don't currently do any writes to their map, but we could easily accidentally reintroduce this bug in the future if a listener did any kind of writes to their map.

I'd suggest either creating a copy per listener so that they can own that data and do what they want with it OR documenting in the contract of the listener interface that implementations MUST NOT modify the AddressSet they receive.

Reply:
Good thinking, I'll go with the former.